[CBR-471] Fix Restoration Performance Regression

KtorZ opened the pull request:
## Description

**/!\** I am still testing this (I had an issue with the very last block application causing the checkpoints to diverge in a non expected way). I believe it's now fixed but I have to wait full restoration. 

We use to perform a db transaction per block which turns out to slow down the process quite a lot.  Instead, we can apply k (.e.g 1000) updates inside one transaction and move on to the next batch. There's a bit of fiddling to do to get things working, but it turns out to be pretty efficient (~600 block/s on my setup, vs 80 block/s without the patch).

The main trick is located in the new `applyHistoricalBlocks` which applies update in sequence, stopping at the first error encountered.

Besides, we should have cought the performance regression earlier. We can't realistically perform restoration as part of our unit or main integration tests because of the time it takes. However, it is perfectly realistic to have it ran nightly, and monitor its output to be inside accepted boundaries.


## Linked issue

[[CBR-471]](https://iohk.myjetbrains.com/youtrack/issue/CBR-471)

## Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply and `~` in the ones that do not: -->
- [x] 🐞 Bug fix (non-breaking change which fixes an issue)
- [ ] 🛠 New feature (non-breaking change which adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🏭 Refactoring that does not change existing functionality but does improve things like code readability, structure etc
- [ ] 🔨 New or improved tests for existing code
- [ ] ⛑ git-flow chore (backport, hotfix, etc)

## Developer checklist
<!--- A mental checklist for the developer submitting the PR. Put an `x` in all the boxes that apply and `~` in the ones that do not: --->
- [x] I have read the [style guide](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/style-guide.md) document, and my code follows the code style of this project.
- [x] If my code deals with exceptions, it follows the [guidelines](https://github.com/input-output-hk/cardano-sl/blob/develop/docs/exceptions.md).
- [x] I have updated any documentation accordingly, if needed. Documentation changes can be reflected in opening a PR on [cardanodocs.com](https://github.com/input-output-hk/cardanodocs.com), amending the inline [Haddock](https://www.haskell.org/haddock/) comments, any relevant README file or one of the document listed in the [docs](https://github.com/input-output-hk/cardano-sl/tree/develop/docs) directory.
- [ ] CHANGELOG entry has been added and is linked to the correct PR on GitHub.

## Testing checklist
<!-- If you aren't providing any tests as part of this PR, use this section to state clearly why. It needs to be a strong motivation and definitely the exception, not the rule. -->
- [x] I have added tests to cover my changes.
- [ ] **TESTS ARE ONGOING All new and existing tests passed.**

## QA Steps
<!--- Which are the steps needed to test this feature, if any? -->

## Screenshots (if available)
<!--- Upload a GIF, an asciinema video or an image demoing the feature -->

dcoutts commented on wallet-new/src/Cardano/Wallet/Kernel/Restore.hs:
```diff
@@ -379,54 +381,50 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
+        let (blocks, txMetas, slotIds) = unzip3 (catMaybes updates)
+        mErr <- update (wallet ^. wallets) (ApplyHistoricalBlocks k rootId blocks)
+        whenLeft mErr (throwM . RestorationApplyHistoricalBlockFailed)
+        forM_ (mconcat txMetas) (putTxMeta (wallet ^. walletMeta))
```
Why is this not another instance of the same problem, with multiple sqlite DB transactions? Is it because this only covers txs that actually belong to our wallet, rather than having an update per-block irrespective?

edsko replied:
I don't think this is a major concern, since most blocks contain no transactions that are relevant to the wallet, and so this won't do anything. But it might indeed make sense to introduce a `putTxMetas` call; it would be trivial to use here.

dcoutts commented on wallet-new/src/Cardano/Wallet/Kernel/Restore.hs:
```diff
@@ -379,54 +381,50 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
     wId :: WalletId
     wId = WalletIdHdRnd rootId
 
     -- Process the restoration of the block with the given 'HeaderHash'.
```
This comment would now be out of date right? If I followed correctly, it's restore _starting from_ the block with the given hash.

KtorZ replied:
Fixed :+1:

dcoutts commented on wallet-new/src/Cardano/Wallet/Kernel/Restore.hs:
```diff
@@ -379,54 +381,50 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
+                    lift (nextHistoricalHash currentHash) >>= \case
+                        Nothing  -> throwM (RestorationFinishUnreachable tgtHash currentHash)
+                        Just hh' -> put (hh', False)
+                lift $ getPrefilteredBlockOrThrow genesisHash currentHash
```
The use of a state monad here seems a bit odd to me. This is just keeping the `currentHash` as the running value right? I think I'd just write this as a local recursive action with the `currentHash` as a local arg.

dcoutts commented on wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs:
```diff
@@ -288,6 +289,17 @@ applyBlock k context restriction blocks = runUpdateDiscardSnapshot $ do
+                      -> Update DB (Either Spec.ApplyBlockFailed ())
+applyHistoricalBlocks k rootId updates =
+    runExceptT $ forM_ updates $ \(context, blocks) ->
+        ExceptT $ applyHistoricalBlock k context rootId blocks
```
Nice.

erikd commented:
tried running the Nix scipt myself and first off, got a few locale related errors:
```
$ nix-build -A acceptanceTests.mainnet.full -o acceptance-tests && ./acceptance-tests
warning: dumping very large path (> 256 MiB); this may run out of memory
/nix/store/73pfw2r6c9g9v5rnyzwg38fm2ap3j8fd-acceptance-tests-mainnet
/nix/store/0pc9pzaqdbygxr2nyng5h71mkwbwv2fc-bash-4.4-p12/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_AU.UTF-8)
./acceptance-tests: line 23: warning: setlocale: LC_ALL: cannot change locale (en_GB.UTF-8): No such file or directory
Launching wallet node: /nix/store/g5m0r33dxw9zjv88zypbbgb437jfixam-wallet-connect-to-mainnet
```
but it did seem to sync:
```
....
12:37:16 Blockchain syncing: 97%  1673651/1710959 blocks
12:37:29 Blockchain syncing: 98%  1684402/1710959 blocks
12:37:40 Blockchain syncing: 98%  1692786/1710959 blocks
12:37:50 Blockchain syncing: 99%  1704754/1710959 blocks
Blockchain synced: 100%  1710959 blocks

Blockchain sync complete!
Blockchain height: 1710959 blocks
Elapsed time: 1846 seconds
```
but then during wallet restore I got:
```
Going to restore a wallet
{"data":{"id":"Ae2tdPwUPEZ1UMga1bnFFD7FoFN9ZFoPLCWLaGcqNnfovcrcT3oMSuAz8DD","name":"Acceptance Wallet","balance":0,"hasSpendingPassword":false,"spendingPasswordLastUpdate":"2018-10-25T01:38:00.112482","createdAt":"2018-10-25T01:38:07.108056","assuranceLevel":"normal","syncState":{"tag":"restoring","data":{"throughput":{"quantity":0,"unit":"blocksPerSecond"},"percentage":{"quantity":0,"unit":"percent"},"estimatedCompletionTime":{"quantity":1711414000,"unit":"milliseconds"}}},"type":"regular"},"status":"success","meta":{"pagination":{"totalPages":1,"page":1,"perPage":1,"totalEntries":1}}}jq: error (at <stdin>:1): Cannot index string with string "data"
jq: error (at <stdin>:1): Cannot index string with string "data"
12:38:07 Wallet restoring...
jq: error (at <stdin>:1): Cannot index string with string "data"
jq: error (at <stdin>:1): Cannot index string with string "data"
12:38:17 Wallet restoring...
erikd at ada in ~ 
```

edsko commented on wallet-new/src/Cardano/Wallet/Kernel/DB/AcidState.hs:
```diff
@@ -814,6 +826,7 @@ makeAcidic ''DB [
     , 'applyBlock
     , 'switchToFork
+    , 'applyHistoricalBlocks
     , 'applyHistoricalBlock
```
Perhaps we should even _remove_ the acidic version of `applyHistoricalBlock`?

adinapoli-iohk replied:
Yes, it looks like a good idea.

KtorZ replied:
removed :+1:

dcoutts replied:
Note that that would break any existing wallet DBs, but since this is not released yet you can get away with it.

KtorZ replied:
Ho! Yes indeed ^.^

edsko commented on wallet-new/src/Cardano/Wallet/Kernel/Restore.hs:
```diff
@@ -379,54 +381,51 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
+
+        -- Flush to Acid-State & SQLite
+        k <- getSecurityParameter (wallet ^. walletNode)
+        let (blocks, txMetas, slotIds) = unzip3 (catMaybes updates)
```
I think I'd rewrite this code slightly; after discussing this code with @KtorZ , the `catMaybes` here has a dual purpose: any EBBs will be `Nothing`, and once we reach the final block halfway the batch, all the elements after that final block will also be `Nothing`. I find the style a bit confusing, and would probably prefer an explicit recursion style with an accumulator. That said, I don't think it's _wrong_ what's here; but I think it does at least need a comment.

adinapoli-iohk commented on scripts/test/acceptance/default.nix:
```diff
@@ -100,4 +100,57 @@ in
+    curl --silent --cacert ${stateDir}/tls/client/ca.crt --cert ${stateDir}/tls/client/client.pem -H "Content-Type: application/json" https://${wallet.walletListen}/api/v1/wallets \
+      -d '{
+      "operation": "restore",
+      "backupPhrase": ["session", "ring", "phone", "arrange", "notice", "gap", "media", "olympic", "water", "road", "spider", "rate"],
```
Hopefully people won't try to use this one for production wallets 😉 

(I always feel like we should have some mechanism for preventing mainnet to accept any of the mnemonics we use in tests etc).

KtorZ replied:
Haha, I thought the same things :no_mouth: ... I hope it's hidden enough in the nix entrails that no one will dig this out... right ? :fearful:

adinapoli-iohk commented on wallet-new/src/Cardano/Wallet/Kernel/Restore.hs:
```diff
@@ -379,54 +381,51 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
         Nothing -> withNode $ getFirstGenesisBlockHash genesisHash
         Just sh -> nextHistoricalHash sh >>= maybe (throwM $ RestorationSuccessorNotFound sh) pure
-    restore genesisHash startingPoint NoTimingData
+    restore genesisHash startingPoint 1000
```
Minor one -- might be nice to explicitly have a `where` clause for this "magic number", to clarify its purpose.

KtorZ commented:
```hs
    getBatch genesisHash hh batchSize = go [] batchSize hh
        where
            go !updates !n !currentHash | n <= 0 =
                return (updates, currentHash)

            go !updates _ !currentHash | currentHash == tgtHash =
                getPrefilteredBlockOrThrow genesisHash currentHash >>= \case
                    Nothing -> go updates 0 currentHash
                    Just u  -> go (updates ++ [u]) 0 currentHash

            go !updates !n !currentHash = do
                nextHash <- nextHistoricalHash currentHash >>= \case
                    Nothing  -> throwM (RestorationFinishUnreachable tgtHash currentHash)
                    Just hh' -> return hh'

                getPrefilteredBlockOrThrow genesisHash currentHash >>= \case
                    Nothing -> go updates (n - 1) nextHash
                    Just u  -> go (updates ++ [u]) (n - 1) nextHash
```

Currently trying out this. I found a nice way to put this. A bit afraid about the boundary case so I'll go through a full restoration once again and see :)

KtorZ commented:
Addressed comments above; went for the algorithm described above with accumulators and recursion instead of the `StateMonad`. I've re-ran the entire script, caught another error in the acceptance bash script (god I hate bash) and fixed it as well.
